pkg/metrics/builder/builder.go

```diff
@@ -0,0 +1,142 @@
+	name string
+	// value stores the value of the metric that needs to be pushed. Collected from the installer.
+	value float64
+	// buckets keeps a list of bucket values that will be used during the creation of a Histogram
+	// object.
+	buckets []float64
+	// metricType defines what type of a collector object should the PromCollector function return.
+	metricType MetricType
+}
+
+// MetricOpts contains the properties that are required to create a MetricBuilder object.
+type MetricOpts struct {
+	// labels contains the list of all label keys that the collector object will have.
+	Labels []string
+	// desc describes the metric that is being created. This field will be shown on Prometheus
+	// dashboard as a help text.
+	Desc string
+	// name describes the name of the metric that needs to be pushed.
+	Name string
+	// buckets keeps a list of bucket values that will be used during the creation of a Histogram
+	// object.
+	Buckets []float64
+	// metricType defines what type of a collector object should the PromCollector function return.
+	MetricType MetricType
+}
+
+// MetricType defines what types of metrics can be created. Restricted by the types of the Prometheus
+// Collector types.
+type MetricType string
+
+const (
+	// Histogram denotes that the type of the collector object should be a Prometheus Histogram.
+	Histogram MetricType = "Histogram"
+	// Counter denotes that the type of the collector object should be a Prometheus Counter.
+	Counter MetricType = "Counter"
+)
+
+// PromCollector function creates the required prometheus collector object with the values
+// it has in the MetricBuilder calling object.
+func (m MetricBuilder) PromCollector() (prometheus.Collector, error) {
+	switch m.metricType {
+	case Counter:
+		return m.buildCounter(), nil
+	case Histogram:
+		return m.buildHistogram(), nil
+	default:
+		return nil, fmt.Errorf(`invalid metric builder type "%s". cannot create collector`, m.metricType)
+	}
+}
+
+// buildCounter returns a prometheus counter object with the value and labels set
+// in the MetricBuilder object.
+func (m *MetricBuilder) buildCounter() prometheus.Collector {
+	collector := prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name:        m.name,
+			Help:        m.desc,
+			ConstLabels: m.labelKeyValues,
+		},
+	)
+	collector.Add(m.value)
+	return collector
+}
+
+// buildHistogram returns a prometheus Histogram object with the value, labels and buckets set
+// in the MetricBuilder object.
+func (m *MetricBuilder) buildHistogram() prometheus.Collector {
+	collector := prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:        m.name,
+			Help:        m.desc,
+			Buckets:     m.buckets,
+			ConstLabels: m.labelKeyValues,
+		},
+	)
+	collector.Observe(m.value)
+	return collector
+}
+
+// NewMetricBuilder creates a new MetricBuilder object with the default values for the field.
+func NewMetricBuilder(opts MetricOpts, value float64, labelKeyValues map[string]string) (*MetricBuilder, error) {
```

Comment:
Perhaps we should add a check if histogram, require buckets.

Reply:
If we don't provide the buckets, the histogram object automatically creates an exponential bucket with default values. I thought maybe it could be used.